### PR TITLE
Add decorator DoRetryForStatusCodes for retry logic. Add CookieJar.

### PR DESCRIPTION
**Pull request by garimakhulbe**
- Add retry logic for 408, 500, 502, 503 and 504 status codes.
- Add CookieJar in Client.
- Change url path and query encoding logic.
- Fix DelayForBackoff for proper exponential delay.

### CODE REVIEW

**Comment by garimakhulbe** [comment]
@ahmetalpbalkan @brendandixon @colemickens Please take a look.

**Comment by ahmetalpbalkan** [comment]
Why are we adding a cookie jar?

**Comment by garimakhulbe** [comment]
@ahmetalpbalkan Autorest code generator has some retry tests which use cookie.

File: autorest/preparer_test.go [hunk 1]
```diff
@@ -70,7 +70,7 @@ func ExampleCreatePreparer() {
 	} else {
 		fmt.Println(r.URL)
 	}
-	// Output: https://microsoft.com/a/b/c
```
Comment (brendandixon) [inline autorest/preparer_test.go]: @garimakhulbe Why did you remove the "/microsoft.com"? The resulting URLs are not valid (they lack the required "//" after the HTTP scheme).

File: autorest/sender.go [hunk 2]
```diff
@@ -222,13 +254,18 @@ func WithLogging(logger *log.Logger) SendDecorator {
...
 
 // DelayForBackoff invokes time.After for the supplied backoff duration raised to the power of
-// passed attempt (i.e., an exponential backoff delay). Backoff may be zero. The delay may be
-// canceled by closing the passed channel. If terminated early, returns false.
+// passed attempt (i.e., an exponential backoff delay). Backoff duration is in seconds and can set
+// to zero for no delay. The delay may be canceled by closing the passed channel. If terminated early,
+// returns false.
 func DelayForBackoff(backoff time.Duration, attempt int, cancel <-chan struct{}) bool {
 	select {
-	case <-time.After(time.Duration(math.Pow(float64(backoff), float64(attempt)))):
```
Comment (brendandixon) [inline autorest/sender.go]: @garimakhulbe Is there a reason you chose to remove the exponential back off?

File: autorest/sender_test.go [hunk 3]
```diff
@@ -144,7 +144,7 @@ func TestSend(t *testing.T) {
 func TestAfterDelayWaits(t *testing.T) {
 	client := mocks.NewSender()
 
-	d := 5 * time.Millisecond
```
Comment (brendandixon) [inline autorest/sender_test.go]: @garimakhulbe I purposely chose Millisecond to keep the tests from getting too long. Were those values not working of you?

File: autorest/utility.go [hunk 4]
```diff
@@ -97,7 +99,63 @@ func ensureValueString(value interface{}) string {
...
+		return pathEscape(s)
+	case "query":
+		return queryEscape(s)
+	default:
+		return s
+	}
+}
+
+func pathEscape(s string) string {
+	return strings.Replace(url.QueryEscape(s), "+", "%20", -1)
```
Comment (brendandixon) [inline autorest/utility.go]: @garimakhulbe Do we need to escape the % in this string?

File: autorest/utility_test.go [hunk 5]
```diff
@@ -186,18 +186,52 @@ func TestEnsureStrings(t *testing.T) {
...
 		"nil":    "",
+		"bytes":  string([]byte{255, 254, 253}),
 	}
 	v := ensureValueStrings(m)
 	if !reflect.DeepEqual(v, r) {
 		t.Fatalf("autorest: ensureValueStrings returned %v\n", v)
 	}
 }
 
+func ExampleStringWithSeparator() {
```
Comment (brendandixon) [inline autorest/utility_test.go]: @garimakhulbe This and the following should be tests. Examples become part of the documentation. I don't believe we need to or should document these methods.

File: autorest/sender.go (hunk 2 above)
Comment (garimakhulbe) [inline autorest/sender.go]:
```Go
    backoff := 2 * time.Second
	fmt.Printf("float64(backoff) : %v\n\n",float64(backoff))
    // float64(backoff) : 2e+09

	for attempt := 0; attempt < 3; attempt++ {
		s := time.Duration(math.Pow(float64(backoff), float64(attempt)))
		fmt.Printf("Attempt # %v --  %v**%v = %v ~ %v \n",
			attempt, float64(backoff), float64(attempt), math.Pow(float64(backoff), float64(attempt)), s)
	}
	// Output :
	// Attempt # 0 --  2e+09**0 = 1 ~ 1ns 
	// Attempt # 1 --  2e+09**1 = 2e+09 ~ 2s 
	// Attempt # 2 --  2e+09**2 = 4e+18 ~ 1111111h6m40s  ---> weird duration

		
	fmt.Printf("\nDuration: %v \n",time.Duration(math.Pow(backoff.Seconds(), float64(3)))) // 
	// Duration: 8ns  --> changed float64(backoff) to backoff.Seconds()) but time still in nanosec
	
	
	fmt.Printf("\nDuration: %v \n",time.Duration(math.Pow(backoff.Seconds(), float64(3)))*time.Second)
	// Duration: 8s --> in seconds ( time.Duration(....)*time.Second )

File: autorest/preparer_test.go (hunk 1 above)
Comment (garimakhulbe) [inline autorest/preparer_test.go]: @brendandixon We are using url.Opaque now and url.String() uses below form if url.Opaque is non-empty.

```scheme:opaque?query#fragment       ---- no host ```

Comment (brendandixon) [inline autorest/preparer_test.go]: @garimakhulbe Host must still be there or the URL is invalid. Opaque only means that the portion of the URI is not being further inspected (as it is scheme dependent). For the http(s) schemes, the "//host.xxx" is required.

Comment (garimakhulbe) [inline autorest/preparer_test.go]: Yes, host is there in URL.

```Go
req := &http.Request{
		URL: &url.URL{
			Host:   "microsoft.com",
			Scheme: "https",
			Opaque: "/a/b/c",
		},
	}
fmt.Println("Host: ", req.URL.Host)
fmt.Println("URL: ", req.URL)

//
// Host:  microsoft.com
// URL:  https:/a/b/c

Comment (brendandixon) [inline autorest/preparer_test.go]: @garimakhulbe Ok. I'm confused. "https:/a/b/c" is __not__ a valid URL. It lacks the required host component.

Comment (colemickens) [inline autorest/preparer_test.go]: Is some of the confusion stemming from the fact that Request has it's own Host field that overrides the URL's Host?

Comment (garimakhulbe) [inline autorest/preparer_test.go]: Removed url.Opaque. It was giving invalid url string. URL string does not fetch host from url.Host when opaque is non-empty and probably expects everything in opaque. Not sure.
```
http:/a/b/c
```

I have updated WithPathParameters() to fix this. Instead of using WithPath() to concatenate path  and then WithPathParameters() for substituting the parameters, we will only use WithPathParameters() when we have path with parameters and it will build the path and parse to url.

```Go
//For URL: http://microsoft.com/a/{b}/c with b = xyz.

Prepare(..
           WithBaseURI("http://microsoft.com")  // --> url : http://microsoft.com
           WithPathParameters("/a/{b}/c",{"b":"xyz"}))   // --> url: http:/microsoft.com/a/xyz/c

//For simple URL with no parameters in path:   http://microsoft.com/m/n

Prepare(..
           WithBaseURI("http://microsoft.com")  // --> url : http://microsoft.com
           WithPath("/m/n")     // --> //url: http:/microsoft.com/m/n
```

File: autorest/sender_test.go (hunk 3 above)
Comment (garimakhulbe) [inline autorest/sender_test.go]: Yes, milliseconds with above change was giving zero seconds delay so increased it.
```Go
backoff := 5* time.Millisecond
fmt.Println(backoff.Seconds())
fmt.Println(time.Duration(backoff.Seconds()))

//Output:
0.005
0

```

File: autorest/utility.go (hunk 4 above)
Comment (garimakhulbe) [inline autorest/utility.go]: We don't need to. We are escaping + to %20 and %20 is valid for path so we don't need to escape %.

File: autorest/utility_test.go (hunk 5 above)
Comment (garimakhulbe) [inline autorest/utility_test.go]: done.

**Comment by garimakhulbe** [comment]
@brendandixon @ahmetalpbalkan Is this PR good for merge?

File: autorest/sender.go [hunk 6]
```diff
@@ -184,6 +186,36 @@ func DoRetryForAttempts(attempts int, backoff time.Duration) SendDecorator {
...
 }
 
+// DoRetryForStatusCodes returns a SendDecorator that retries for specified statusCodes for up to the specified
+// number of attempts, exponentially backing off between requests using the supplied backoff
+// time.Duration (which may be zero). Retrying may be canceled by closing the optional channel on
+// the http.Request.
+func DoRetryForStatusCodes(attempts int, backoff time.Duration, codes ...int) SendDecorator {
+	return func(s Sender) Sender {
+		return SenderFunc(func(r *http.Request) (resp *http.Response, err error) {
+			b := []byte{}
```
Comment (ahmetalpbalkan) [inline autorest/sender.go]: `var b []byte{}` would be better (1 less alloc)

File: autorest/sender.go [hunk 7]
```diff
@@ -184,6 +186,36 @@ func DoRetryForAttempts(attempts int, backoff time.Duration) SendDecorator {
...
+			if r.Body != nil {
+				b, err = ioutil.ReadAll(r.Body)
+				if err != nil {
+					return resp, err
```
Comment (ahmetalpbalkan) [inline autorest/sender.go]: try wrapping the error with a message.

File: autorest/sender.go [hunk 8]
```diff
@@ -184,6 +186,36 @@ func DoRetryForAttempts(attempts int, backoff time.Duration) SendDecorator {
...
+			attempts++
+			for attempt := 0; attempt < attempts; attempt++ {
+				r.Body = ioutil.NopCloser(bytes.NewBuffer(b))
+				resp, err = s.Do(r)
+				if err != nil || !ResponseHasStatusCode(resp, codes...) {
+					return resp, err
+				}
+				DelayForBackoff(backoff, attempt, r.Cancel)
+			}
+			return resp, err
```
Comment (ahmetalpbalkan) [inline autorest/sender.go]: shouldn't this error be wrapped to indicate that we had retries and this request has exceeded retries

File: autorest/sender.go [hunk 9]
```diff
@@ -222,11 +254,12 @@ func WithLogging(logger *log.Logger) SendDecorator {
...
 // DelayForBackoff invokes time.After for the supplied backoff duration raised to the power of
-// passed attempt (i.e., an exponential backoff delay). Backoff may be zero. The delay may be
-// canceled by closing the passed channel. If terminated early, returns false.
+// passed attempt (i.e., an exponential backoff delay). Backoff duration is in seconds and can set
+// to zero for no delay. The delay may be canceled by closing the passed channel. If terminated early,
+// returns false.
 func DelayForBackoff(backoff time.Duration, attempt int, cancel <-chan struct{}) bool {
 	select {
-	case <-time.After(time.Duration(math.Pow(float64(backoff), float64(attempt)))):
+	case <-time.After(time.Duration(math.Pow(backoff.Seconds(), float64(attempt))) * time.Second):
```
Comment (ahmetalpbalkan) [inline autorest/sender.go]: .Seconds() essentially takes your ability to use sub-second backoff durations. If I give this a 4.5s, it will use 4. If I give it a 0.8 it will always be 0.

My suggestion, use `.Milliseconds()` to get integers and use them in math.Pow() and then multiply it with `time.Milliseconds` and pass it to `time.After`.

Comment (brendandixon) [inline autorest/sender.go]: @garimakhulbe @ahmetalpbalkan I'm not certain it matters much. Sub-second back off seldom makes sense.

Comment (ahmetalpbalkan) [inline autorest/sender.go]: fair. I'd implement this with `backoff * (2^attempt)`. I don't think `backoff^attempt` is a very good idea, for 10s, it will go off quite exponentially.

Comment (brendandixon) [inline autorest/sender.go]: @ahmetalpbalkan The idea behind an exponential is that, if timeouts occur for our process, then they're occurring for others. Exponentially waiting spreads those request farther out, faster over time to give the underlying process a chance to recover. Simple linear back off tends to keep requests packed close together, often exacerbating the problem.

Comment (ahmetalpbalkan) [inline autorest/sender.go]: @brendandixon what I suggested above is still exponential.

Comment (garimakhulbe) [inline autorest/sender.go]: Backoff*(2^attempt) looks good. Its exponential and won't increase steeply. We can switch to this.

backoff^attempt for 10 seconds
10, 100, 1000,..

whereas backoff * (2^attempt)
10, 20, 40, 80,..

@brendandixon Is it look fine?

Comment (brendandixon) [inline autorest/sender.go]: @garimakhulbe That will work well.

Comment (ahmetalpbalkan) [inline autorest/sender.go]: @garimakhulbe Great, thanks.

For more see "Example exponential back-off algorithm" in https://en.wikipedia.org/wiki/Exponential_backoff
or https://www.awsarchitectureblog.com/2015/03/backoff.html

**Comment by garimakhulbe** [comment]
Backoff delay fixed. Let me know if overall LGTY. I will send small PR from now on. :)

**Comment by ahmetalpbalkan** [comment]
LGTM
